### PR TITLE
[Dygraph]Integration sharding stage2 function

### PR DESCRIPTION
**Pull request by Baibaifan**
### PR types

Performance optimization 
### PR changes

Others
### Describe

Integration sharding stage2 function
1.Support group = None
2.Support param_groups for optimizer

```python3
import paddle
from paddle.distributed.fleet.meta_optimizers.dygraph_optimizer.sharding_optimizer_stage2 import ShardingOptimizerStage2
from paddle.distributed.fleet.meta_parallel.sharding.sharding_stage2 import ShardingStage2

fleet.init(is_collective=True)
group = paddle.distributed.new_group([0, 1])

# wrap model & optimizer 
model = model_class(...)
oss_optimizer = ShardingOptimizer(params=model.parameters(), optim=optimizer, group=group)
model = ShardingStage2(model, oss_optimizer, group=group)

# use optimizer as normal
img, label = data
label.stop_gradient = True
img.stop_gradient = True
out = model(img)

loss = paddle.nn.functional.cross_entropy(input=out, label=label)
oss_optimizer.step()
oss_optimizer.clear_grad()
```
image

### CODE REVIEW

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/sharding_optimizer_stage2.py [hunk 1]
```diff
@@ -78,13 +75,16 @@ def __init__(self,
...
-        self._segment_params = []
+        self.__segment_params = []
         self._rank_buffer_size = {}  # {dtype: {rank: numel+alignment}}
         self._param2align = {}  # {param.name: align}
 
         # Default information
         self._optim_defaults = kw
         self._optim = optim
+        self._ori_parameter_list = copy.deepcopy(self._optim._parameter_list)
+        self._ori_param_groups = copy.deepcopy(self._optim._param_groups)
```
Comment (ForFishes) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/sharding_optimizer_stage2.py]: deepcopy increase memory..

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/sharding_optimizer_stage2.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def __init__(self,
                 filter(lambda x: x.trainable and x.dtype == Type.fp16.value,
                        self._local_params))) > 0
 
-        assert group is not None, "Distributed communication group is must be gived"
+        assert group is not None, "Distributed communication group is must be given"
```
Comment (ForFishes) [inline python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/sharding_optimizer_stage2.py]: need support global group if group=None

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/sharding_optimizer_stage2.py (hunk 1 above)
Comment (Baibaifan) [reply python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/sharding_optimizer_stage2.py]: 已修复，改成引用传递。

File: python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/sharding_optimizer_stage2.py (hunk 2 above)
Comment (Baibaifan) [reply python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/sharding_optimizer_stage2.py:97]: 如果取默认_get_global_group()函数，那么group.ranks是个空[]，在broastcast会检查 assert gsrc >= 0, ("src rank out of group, need global rank")，所以现在不支持group为None情况。

Comment (Baibaifan) [reply python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/sharding_optimizer_stage2.py]: 已支持

**Review by ForFishes — DISMISSED** [review]
LGTM

**Review by ForFishes — APPROVED** [review]
LGTM

**Pull request merged**